### PR TITLE
Divergence and Gradient integrators as well as PA versions

### LINKED ISSUE #1158: Ownership of bilinear forms
I have a class where I need to reassemble lots of different Bilinear forms for every iteration/time step. Due to memory, I cannot keep the fully assembled Bilinearforms and Hypre matrices, but I do want to keep the underlying integrators. E.g., loop over the following:
```
        // Form the bilinear form for one index
        ParBilinearForm *blform = new ParBilinearForm(pfes_);
        blform -> AddDomainIntegrator(mass_integrator_);
        blform -> AddDomainIntegrator(transpose_integrators_[i]);
        blform -> AddInteriorFaceIntegrator(trace_integrators_[i]);
        blform -> AddBdrFaceIntegrator(trace_integrators_[i]);
        blform -> Assemble();
        blform -> Finalize();
```
For memory I need to delete the bilinear form when it goes out of scope (or just not use a pointer), but then it deletes all of my class variable integrators `mass_integrator_`, `transpose_integrators_`, and `trace_integrators_`. This ownership appears to be set by `extern_bfs`, but that variable is protected and there does not appear to be a way to modify it. 

This is related to https://github.com/mfem/mfem/issues/793 -- it would be very helpful if the user had some way to access ownership for variables within BilinearForms. 

I'm not sure the best option. I added a simple public function to BilinearForm.cpp
```
   /** Call to avoid deleting external parbilinear forms. */ 
   void KeepExternalBfs() {extern_bfs = true; }
```
but this must account for all added forms. Another option is for each integrator you add there is an option boolean that defaults to True on whether to assume ownership.

### PR DESCRIPTION
**Pull request by jandrej**
This PR introduces the divergence and gradient bilinear integrators. It introduces facilities to apply boundary conditions to general rectangular operators. These changes required a lot of additions to hypre functions, but do not change any existing interfaces.

For PA it was necessary to introduce a `MixedBilinearFormExtension`.

Some application kernels contain shared memory version from @camierjs.

It also includes unit tests for the PA version of the operators.

Resolves #1158.

- [x] Code builds.
- [x] Code passes make style.
- [x] Update CHANGELOG.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/1164 | @jandrej | @tzanio | @mlstowell + @jamiebramwell | 12/2/19 | 01/08/20 | ⌛due 01/20/20 |

### CODE REVIEW

File: fem/bilinearform.hpp [hunk 1]
```diff
@@ -525,6 +531,9 @@ class BilinearForm : public Matrix
    /// Sets diagonal policy used upon construction of the linear system
    void SetDiagonalPolicy(DiagonalPolicy policy);
 
+   /// Indicate that integrators are not owned by the BilinearForm
+   void SetExternBFS(int e = 1) { extern_bfs = e; };
```
Comment (tzanio) [inline fem/bilinearform.hpp]: I will suggest a move descriptive name and dropping the parameter, e.g. `void UseExternalIntegrators()'.

File: fem/bilinearform.cpp [hunk 2]
```diff
@@ -1470,17 +1561,91 @@ void MixedBilinearForm::EliminateTestDofs (const Array<int> &bdr_attr_is_ess)
...
+      ext->FormRectangularSystemOperator(trial_tdof_list, test_tdof_list, A);
+      return;
+   }
+
+   const SparseMatrix *test_P = test_fes->GetConformingProlongation();
+   const SparseMatrix *trial_P = trial_fes->GetConformingProlongation();
+
+   mat->Finalize();
+
+   if (test_P) // TODO: Must actually check for trial_P too
```
Comment (pazner) [inline fem/bilinearform.cpp]: This TODO should be addressed?

File: fem/bilinearform.cpp [hunk 3]
```diff
@@ -1470,17 +1561,91 @@ void MixedBilinearForm::EliminateTestDofs (const Array<int> &bdr_attr_is_ess)
...
+
+   Array<int> ess_trial_tdof_marker, ess_test_tdof_marker;
+   FiniteElementSpace::ListToMarker(trial_tdof_list, trial_fes->GetTrueVSize(),
+                                    ess_trial_tdof_marker);
+   FiniteElementSpace::ListToMarker(test_tdof_list, test_fes->GetTrueVSize(),
+                                    ess_test_tdof_marker);
+
+   mat_e = new SparseMatrix(mat->Height(), mat->Width());
+   mat->EliminateCols(ess_trial_tdof_marker, *mat_e);
+   // TODO: WP: are we doing the right thing here?
```
Comment (pazner) [inline fem/bilinearform.cpp]: This TODO can be removed

File: fem/bilinearform.hpp (hunk 1 above)
Comment (pazner) [reply fem/bilinearform.hpp]: +1 for a more descriptive name

File: fem/bilinearform_ext.hpp [hunk 4]
```diff
@@ -134,6 +136,120 @@ class MFBilinearFormExtension : public BilinearFormExtension
...
+   {}
+   void FormRectangularLinearSystem(const Array<int> &trial_tdof_list,
+                                    const Array<int> &test_tdof_list,
+                                    Vector &x, Vector &b,
+                                    OperatorHandle &A, Vector &X, Vector &B) {}
+   void Mult(const Vector &x, Vector &y) const {}
+   void MultTranspose(const Vector &x, Vector &y) const {}
+   void Update() {}
+   ~FAMixedBilinearFormExtension() {}
+};
```
Comment (pazner) [inline fem/bilinearform_ext.hpp]: Should we hold off on this FA stuff? It's more-or-less a placeholder now

File: fem/bilininteg.cpp [hunk 5]
```diff
@@ -1925,7 +1998,7 @@ void VectorFEMassIntegrator::AssembleElementMatrix2(
    }
 }
 
-void VectorDivergenceIntegrator::AssembleElementMatrix2(
+void DivergenceIntegrator::AssembleElementMatrix2(
```
Comment (pazner) [inline fem/bilininteg.cpp]: This name change will break backwards compatibility (although I agree it's a better name)

File: linalg/operator.cpp [hunk 6]
```diff
@@ -336,4 +412,76 @@ void ConstrainedOperator::Mult(const Vector &x, Vector &y) const
...
+   const Array<int> &trial_list,
+   const Array<int> &test_list,
+   bool _own_A)
+   : Operator(A->Height(), A->Width()), A(A), own_A(_own_A)
+{
+   // 'mem_class' should work with A->Mult() and MFEM_FORALL():
+   mem_class = A->GetMemoryClass()*Device::GetMemoryClass();
+   MemoryType mem_type = GetMemoryType(mem_class);
+   trial_list.Read(); // TODO: just ensure 'list' is registered, no need to copy it
+   test_list.Read(); // TODO: just ensure 'list' is registered, no need to copy it
```
Comment (pazner) [inline linalg/operator.cpp]: Have these TODOs been addressed?

File: linalg/operator.hpp [hunk 7]
```diff
@@ -30,7 +31,21 @@ class Operator
...
+
+   /// Returns RAP Operator of this, taking in input/output Prolongation matrices
+   Operator *SetupRAP(const Operator *Pi, const Operator *Po);
+
 public:
+   /// Initializes memory for true vectors of linear system
+   void InitTVectors(const Operator *Po, const Operator *Ri,
+                     Vector &x, Vector &b,
+                     Vector &X, Vector &B) const;
+
```
Comment (pazner) [inline linalg/operator.hpp]: Should this be in the public interface? Maybe it can be moved to be private.

File: fem/bilininteg_divergence.cpp [hunk 8]
```diff
@@ -0,0 +1,1073 @@
...
+      double (*DQQ2)[MQ1][MQ1] = (double (*)[MQ1][MQ1]) (sm1+2);
+      double (*QQQ0)[MQ1][MQ1] = (double (*)[MQ1][MQ1]) (sm0+0);
+      double (*QQQ1)[MQ1][MQ1] = (double (*)[MQ1][MQ1]) (sm0+1);
+      double (*QQQ2)[MQ1][MQ1] = (double (*)[MQ1][MQ1]) (sm0+2);
+      double (*QQD0)[MQ1][MD1] = (double (*)[MQ1][MD1]) (sm1+0);
+      //double (*QQD1)[MQ1][MD1] = (double (*)[MQ1][MD1]) (sm1+1);
+      //double (*QQD2)[MQ1][MD1] = (double (*)[MQ1][MD1]) (sm1+2);
+      double (*QDD0)[MD1][MD1] = (double (*)[MD1][MD1]) (sm0+0);
+      //double (*QDD1)[MD1][MD1] = (double (*)[MD1][MD1]) (sm0+1);
+      //double (*QDD2)[MD1][MD1] = (double (*)[MD1][MD1]) (sm0+2);
```
Comment (pazner) [inline fem/bilininteg_divergence.cpp]: These commented-out lines can be removed

File: fem/bilininteg_gradient.cpp [hunk 9]
```diff
@@ -0,0 +1,800 @@
...
+                                  const Vector &_op,
+                                  const Vector &_x,
+                                  Vector &_y,
+                                  const int tr_d1d = 0,
+                                  const int te_d1d = 0,
+                                  const int q1d = 0)
+{
+   // TODO
+   MFEM_ASSERT(false, "SHARED MEM NOT PROGRAMMED YET");
+}
```
Comment (pazner) [inline fem/bilininteg_gradient.cpp]: Better to remove the placeholder for the shared memory version until it's implemented

File: fem/bilininteg_gradient.cpp [hunk 10]
```diff
@@ -0,0 +1,800 @@
...
+                                       const Array<double> &b,
+                                       const Vector &_op,
+                                       const Vector &_x,
+                                       Vector &_y,
+                                       int tr_d1d = 0,
+                                       int te_d1d = 0,
+                                       int q1d = 0)
+{
+   // TODO
+   MFEM_ASSERT(false, "Gradient PA Apply Transpose 3D NOT PROGRAMMED YET");
```
Comment (pazner) [inline fem/bilininteg_gradient.cpp]: Nicer error message? e.g. `"GradientPAApplyTranspose 3D not implemented"`

File: fem/bilininteg_gradient.cpp [hunk 11]
```diff
@@ -0,0 +1,800 @@
...
+   if (dim == 3)
+   {
+      if (transpose)
+      {
+         return PAGradientApplyTranspose3D(NE,B,G,Bt,op,x,y,TR_D1D,TE_D1D,Q1D);
+      }
+      else
+      {
+         return PAGradientApply3D(NE,B,G,Bt,op,x,y,TR_D1D,TE_D1D,Q1D);
+      }
```
Comment (pazner) [inline fem/bilininteg_gradient.cpp]: Why do we check for transpose in the `dim=3` case but not `dim=2`? And is the `transpose` flag ever set given that `GradientIntegrator::AddMultTransposePA` (defined below) throws an error?

File: fem/bilininteg_divergence.cpp [hunk 12]
```diff
@@ -0,0 +1,1073 @@
...
+                              bool transpose=false)
+{
+   if (dim == 2)
+   {
+      return PADivergenceApply2D(NE,B,G,Bt,op,x,y,TR_D1D,TE_D1D,Q1D);
+   }
+   if (dim == 3)
+   {
+      return PADivergenceApply3D(NE,B,G,Bt,op,x,y,TR_D1D,TE_D1D,Q1D);
+   }
```
Comment (pazner) [inline fem/bilininteg_divergence.cpp]: @tzanio @v-dobrev @camierjs We need to figure out a way of doing template specialization for common cases for these kernels without hugely increasing compile times. Here we're only using the default fallback kernel because otherwise (now that we're adding more and more kernels) the compilation is just too long. Also the `switch ((D1D << 4 ) | Q1D)` approach is error-prone and requires too much manual intervention (e.g. picking exactly the right Q1D to match the quadrature rule).

File: fem/bilinearform.cpp (hunk 2 above)
Comment (jandrej) [reply fem/bilinearform.cpp]: I'm not 100% sure what we should do here.

File: linalg/hypre_parcsr.hpp [hunk 13]
```diff
@@ -45,7 +45,14 @@ void hypre_ParCSRMatrixEliminateAXB(hypre_ParCSRMatrix *A,
 void hypre_ParCSRMatrixEliminateAAe(hypre_ParCSRMatrix *A,
                                     hypre_ParCSRMatrix **Ae,
                                     HYPRE_Int num_rowscols_to_elim,
-                                    HYPRE_Int *rowscols_to_elim);
+                                    HYPRE_Int *rowscols_to_elim,
+                                    int ignore_rows = 0);
+
+/** Eliminate rows from a hypre ParCSRMatrix, setting all entries in the listed
+    rows of the matrix to zero. */
+void hypre_ParCSRMatrixEliminateRows(hypre_ParCSRMatrix *A,
```
Comment (jamiebramwell) [inline linalg/hypre_parcsr.hpp]: This goes beyond this PR, but has anyone looked at upstreaming these functions to HYPRE? Any thoughts @tzanio ?

File: linalg/hypre.hpp [hunk 14]
```diff
@@ -525,6 +525,13 @@ class HypreParMatrix : public Operator
        Ae sum to the original matrix. */
    HypreParMatrix* EliminateRowsCols(const Array<int> &rows_cols);
 
+   /** Eliminate columns from the matrix and store the eliminated
+       elements in a new matrix Ae (returned) so that the modified
+       matrix and Ae sum to the original matrix. */
+   HypreParMatrix* EliminateCols(const Array<int> &cols);
+
+   void EliminateRows(const Array<int> &rows);
```
Comment (jamiebramwell) [inline linalg/hypre.hpp]: This is missing a function description. It's pretty obvious what it does, but I think it's nice to include for completeness.

File: fem/bilininteg_gradient.cpp [hunk 15]
```diff
@@ -0,0 +1,775 @@
+// Copyright (c) 2019, Lawrence Livermore National Security, LLC. Produced at
```
Comment (jamiebramwell) [inline fem/bilininteg_gradient.cpp]: While this is not unique to this PR, these kernels are very hard to read without any comments. As someone who will probably have to write specialized kernels in the future, comments outlining the general method for each operator would be greatly appreciated.

Comment (mlstowell) [reply fem/bilininteg_gradient.cpp]: I have to second this.  Since these files don't have headers we should put detailed, header-level, documentation here before each function.  At a minimum we should have descriptions of the arguments but, as @jamiebramwell suggests, an overview of how the function works would be ideal.

File: tests/unit/fem/test_pa_kernels.cpp [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+   lf.AddDomainIntegrator(new VectorDomainLFIntegrator(fcoeff2));
+   lf.Assemble();
+   field2 -= lf;
+
+   delete mesh;
+
+   return field2.Norml2();
+}
+
+TEST_CASE("PA Divergence", "[PartialAssembly], [MixedBilinearFormBC]")
```
Comment (mlstowell) [inline tests/unit/fem/test_pa_kernels.cpp]: I like your tests but I'm curious about "BC" in "MixedBilinearFormBC".  What does this signify?

Comment (jandrej) [reply tests/unit/fem/test_pa_kernels.cpp]: This is cruft that I didn't clean up. I wanted to test the boundary condition application as well but didn't complete it in the commit.

File: fem/bilinearform.hpp [hunk 17]
```diff
@@ -676,8 +689,25 @@ class MixedBilinearForm : public Matrix
...
    void Assemble(int skip_zeros = 1);
 
+   /// Get the input finite element space prolongation matrix
+   virtual const Operator *GetProlongation() const
+   { return trial_fes->GetProlongationMatrix(); }
+   /// Get the input finite element space restriction matrix
+   virtual const Operator *GetRestriction() const
+   { return trial_fes->GetRestrictionMatrix(); }
+   /// Get the output finite element space prolongation matrix
+   virtual const Operator *GetOutputProlongation() const
```
Comment (mlstowell) [inline fem/bilinearform.hpp]: Since this is an FEM class rather than a linear algebra class perhaps it would be better to use "Trial" and "Test" rather than "" and "Output" in these method names.  Personally, I tend to get "trial" and "test" confused but they are part of the convention.

File: fem/bilinearform.cpp [hunk 18]
```diff
@@ -960,6 +958,18 @@ void BilinearForm::EliminateVDofsInRHS(
    mat->PartMult(vdofs, x, b);
 }
 
+void BilinearForm::Mult(const Vector &x, Vector &y) const
```
Comment (mlstowell) [inline fem/bilinearform.cpp]: Don't we need similar changes to `BilinearForm::AddMult()` to support `ext`?

File: fem/bilinearform.hpp [hunk 19]
```diff
@@ -676,8 +689,25 @@ class MixedBilinearForm : public Matrix
 
    void operator=(const double a) { *mat = a; }
```
Comment (mlstowell) [inline fem/bilinearform.hpp]: I'm a little concerned about methods like this, and the `SpMat` methods earlier, that will fail if partial assembly was chosen.  Certainly for the `SpMat` methods we could make the argument that responsible developers should never call them if they requested that matrices not be assembled but we have no control over this and it can easily lead to runtime errors.  I prefer compile time errors whenever possible.

I wonder what people's reaction would be if I suggested that we should have an abstract base class for bilinear forms which contained a minimal interface and then two or more derived classes for concrete implementations of fully assembled and partially assembled bilinear forms.  Each concrete class could then have separate APIs containing only those methods which could actually be implemented.

File: fem/bilininteg_gradient.cpp [hunk 20]
```diff
@@ -0,0 +1,775 @@
...
+}
+
+// PA Gradient Apply kernel
+void GradientIntegrator::AddMultTransposePA(const Vector &x, Vector &y) const
+{
+   MFEM_ABORT("PA Gradient AddMultTransposePA not yet programmed!");
+}
+
+} // namespace mfem
+
```
Comment (mlstowell) [inline fem/bilininteg_gradient.cpp]: These kernel files are very cleanly written!  Adding comment blocks as described earlier would make them perfect.

File: fem/bilinearform.cpp (hunk 18 above)
Comment (jandrej) [reply fem/bilinearform.cpp]: Do you want me to add them? We did this in the Mixed case because they were needed for some operations.

File: fem/bilinearform.hpp (hunk 19 above)
Comment (jandrej) [reply fem/bilinearform.hpp]: Do you think this should be addressed in this PR or discussed in the meeting for another PR?

Comment (mlstowell) [reply fem/bilinearform.hpp]: A meeting would probably be best.

File: fem/bilinearform.cpp (hunk 18 above)
Comment (mlstowell) [reply fem/bilinearform.cpp]: I would suggests adding them because it doesn't seem like much effort.  However, I won't object if you postpone this and start an issue so we won't forget to address it.

Comment (jandrej) [reply fem/bilinearform.cpp]: I'd prefer to postpone this to a different PR.

**Comment by jandrej** [comment]
I addressed all comments either with changes or replies. I did not address the comments in the kernels, I completely understand the issue with writing them and implementations can lead as an example for people who want customized kernels. As stated in other PR's we have to think harder about how to reduce the labor for this.

File: fem/bilinearform.hpp [hunk 21]
```diff
@@ -325,6 +325,12 @@ class BilinearForm : public Matrix
    /// Get the finite element space restriction matrix
    virtual const Operator *GetRestriction() const
    { return fes->GetConformingRestriction(); }
+   /// Get the output finite element space prolongation matrix
+   virtual const Operator *GetOutputProlongation() const
+   { return GetProlongation(); }
+   /// Get the output finite element space restriction matrix
+   virtual const Operator *GetOutputRestriction() const
```
Comment (mlstowell) [inline fem/bilinearform.hpp]: Should change these from "Output" to "Test" as well or eliminate them.

Comment (jandrej) [reply fem/bilinearform.hpp]: Removed.

**Comment by tzanio** [comment]
Reminder: the reviews for this PR are due in less than a week

cc: @mlstowell, @jamiebramwell

**Comment by tzanio** [comment]
@jamiebramwell were you comments addressed, and if so, can you formally approve?

File: tests/unit/fem/test_pa_kernels.cpp [hunk 22]
```diff
@@ -0,0 +1,227 @@
+#include "catch.hpp"
```
Comment (tzanio) [inline tests/unit/fem/test_pa_kernels.cpp]: Please add the Copyright header in the beginning of this file (as in the other tests).

**Comment by tzanio** [comment]
@jandrej, before we merge this into `next`, can you please do the following:
- [ ] Update `CHANGELOG`
- [ ] Go through [all the comments](https://github.com/mfem/mfem/pull/1164/files), and either resolve them (if fixed) or address them (if not)
- [ ] Stop by to discuss in person if there are questions :)

**Comment by tzanio** [comment]
Merged in `next` for testing ...

File: fem/bilininteg.hpp [hunk 23]
```diff
@@ -43,6 +43,9 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
    /** The result of the partial assembly is stored internally so that it can be
        used later in the methods AddMultPA() and AddMultTransposePA(). */
    virtual void AssemblePA(const FiniteElementSpace &fes);
+   /** Used with BilinearFormIntegrators that have different spaces. */
+   virtual void AssemblePA(const FiniteElementSpace &trial_fes,
+                           const FiniteElementSpace &test_fes);
```
Comment (v-dobrev) [inline fem/bilininteg.hpp]: While building the `next` branch with CUDA enabled, I noticed that this new overload of `AssemblePA` causes numerous warnings like this:
```console
fem/bilininteg.hpp(1671): warning: overloaded virtual function "mfem::BilinearFormIntegrator::AssemblePA" is only partially overridden in class "mfem::GradientIntegrator"
```
One way to fix this is to add `using BilinearFormIntegrator::AssemblePA;` to all subclasses. Another is to change the name of the method slightly, e.g. `MixedAssemblePA`.

Comment (v-dobrev) [reply fem/bilininteg.hpp]: I see that there was a similar issue in #1159 that @camierjs fixed there by adding `using` statements. In that PR the same method as above was introduced at the level of the parent class, `NonlinearFormIntegrator`.

Comment (jandrej) [reply fem/bilininteg.hpp]: I don't have a cuda compiler setup. Can someone who has make the appropriate changes?

Comment (tzanio) [reply fem/bilininteg.hpp]: @camierjs, do you think you can help @jandrej with this?

Comment (camierjs) [reply fem/bilininteg.hpp]: `done`

Comment (jandrej) [reply fem/bilininteg.hpp]: Thanks!

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
@jandrej, can you resolve the conflicts?

**Comment by tzanio** [comment]
Added few more days of testing due to changes...

**Comment by jandrej** [comment]
> @jandrej, can you resolve the conflicts?

Done.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by ben-s-southworth** [comment]
From https://github.com/mfem/mfem/issues/1158 and originally https://github.com/mfem/mfem/issues/793, I don't see anything here that addresses the HypreParMatrix ownership? This comes up more often for me than the integrators, I need a matrix, but not the integrators or bilinear forms.

File: fem/bilinearform.hpp [hunk 24]
```diff
@@ -533,6 +539,9 @@ class BilinearForm : public Matrix
    /// Sets diagonal policy used upon construction of the linear system
    void SetDiagonalPolicy(DiagonalPolicy policy);
 
+   /// Indicate that integrators are not owned by the BilinearForm
+   void UseExternalIntegrators() { extern_bfs = 1; };
```
Comment (v-dobrev) [inline fem/bilinearform.hpp]: @ben-s-southworth, I think this is the method you are looking for.

Comment (v-dobrev) [reply fem/bilinearform.hpp]: At least with respect to #1158.

Comment (ben-s-southworth) [reply fem/bilinearform.hpp]: @v-dobrev Yep, https://github.com/mfem/mfem/issues/1158 is addressed here, but I don't think anything addresses the HypreParMatrices in https://github.com/mfem/mfem/issues/793, unless I missed it

Comment (v-dobrev) [reply fem/bilinearform.hpp]: My understanding is that the discussion in #793 is still open. I don't think it was moved to #1158 -- at least that is my understanding.

In any case, I agree that this PR does not address #793.

Comment (tzanio) [reply fem/bilinearform.hpp]: Correct, the discussion in #793 is outstanding and not addressed here.

File: linalg/hypre_parcsr.hpp (hunk 13 above)
Comment (tzanio) [reply linalg/hypre_parcsr.hpp]: This is on my todo list...

File: linalg/sparsemat.hpp [hunk 25]
```diff
@@ -424,7 +424,7 @@ class SparseMatrix : public AbstractSparseMatrix
    /// Returns whether or not CSR format has been finalized.
    bool Finalized() const { return !A.Empty(); }
    /// Returns whether or not the columns are sorted.
-   bool areColumnsSorted() const { return isSorted; }
+   bool ColumnsAreSorted() const { return isSorted; }
```
Comment (v-dobrev) [inline linalg/sparsemat.hpp]: @tzanio, please do not sneak in interface changes like this -- just because you didn't like the name, and right before merging in master.

Comment (tzanio) [reply linalg/sparsemat.hpp]: This was an editorial decision that I discussed with @jandrej in person. It was not a question of taste, but of conformity with the current naming practice in MFEM.
